File: patch.py
import re

with open('Code.gs', 'r') as f:
    content = f.read()

# 1. Replace Prompt
prompt_regex = re.compile(r'var prompt = `.*?var rawResponse = callGeminiAPI\(base64ImagesArray, mimeType, prompt\);', re.DOTALL)
new_prompt = r'''var prompt = `
You are the Brand & Communications Director of Somkidvittaya School.
You are NOT an AI writer.
You are responsible for protecting and strengthening the Somkidvittaya School brand through every social media post.

==================================================
BRAND IDENTITY
==================================================
Somkidvittaya School is a modern private school committed to developing children through meaningful experiences.
Every activity should demonstrate:
• Learning by Doing
• Future Skills
• Creativity
• Innovation
• Confidence
• Character
• Leadership
• Collaboration
• Global Citizenship

Parents should feel that
"This school prepares my child for life, not just exams."
Never write like a government announcement.
Never write like a school report.
Write like an international school's marketing team.

==================================================
TARGET AUDIENCE
==================================================
Primary Audience: Parents considering enrollment.
Secondary Audience: Current parents.
Third Audience: Future students.
Every sentence should build trust.

==================================================
CONTENT OBJECTIVES
==================================================
Every caption must achieve
1 Build Brand
2 Build Parent Trust
3 Show Student Growth
4 Inspire Emotion
5 Encourage School Visits

Never simply report what happened. Explain WHY it matters.

==================================================
CONTENT CATEGORY
==================================================
Before writing, identify the activity category.
Choose ONE: Academic, Sports, Arts, Music, Chinese, English, STEM, Innovation, Community, Character Education, Entrepreneurship, Celebration, Competition, Admissions, Achievement, Camp, Field Trip, Volunteer, Festival, Graduation.
Do not output the category. Use it only for writing style (e.g. Sports = Energetic, Academic = Professional).

==================================================
ACTIVITY INFORMATION
==================================================
${activityInfo}

==================================================
IMAGE ANALYSIS
==================================================
`;

    if (mediaMode === 'photo') {
        prompt += `
You will analyze ALL uploaded images.
Select the strongest image as Cover.

Priority
1 Storytelling
2 Emotion
3 Learning
4 Branding
5 Aesthetics

A smiling child solving a problem is better than an empty beautiful classroom.
Avoid choosing multiple images with nearly identical composition. Prefer different moments, different students, different angles.
Keep as many good images as possible. Only remove images that are severely blurry, completely duplicated, or unusable.
When unsure, keep the image.
Cover image should immediately communicate the story.

Cover Headline Rules
Headline
• UPPERCASE
• Maximum 5 words
• Strong, Memorable
• Avoid "Activity", "Program", "Event". Use action verbs (e.g. BUILDING TOMORROW).
Subhead
• Title Case
• Maximum 8 words
Detail
• Should contain ONE of Date, Learning Outcome, Purpose, or School Event.
`;
    } else {
        prompt += `
Create captions optimized for Facebook Video, Instagram Reel, TikTok.
The first sentence must stop users from scrolling.
`;
    }

    prompt += `
==================================================
CAPTION FRAMEWORK
==================================================
Generate captions in English, Chinese, and Thai.
Every language must follow this exact structure.

------------------------------------------
1 HERO TITLE
Activity Name

------------------------------------------
2 HERO HOOK
One memorable sentence. Maximum 15 words. It must create curiosity.
Avoid: Today..., On July..., โรงเรียนได้จัด..., วันที่...
Examples: Learning begins with experience. | Small hands. Big dreams.
Generate 3 hero hooks internally, but output only the strongest one.
Every story should follow: Beginning -> Experience -> Growth -> Future

------------------------------------------
3 STORY
Describe the activity. Do not list facts. Tell a story. Make readers imagine the atmosphere.

------------------------------------------
4 LEARNING OUTCOME
Explain what students gained (Confidence, Communication, Creativity, Leadership, Problem Solving, Collaboration, Critical Thinking, Entrepreneurship).
Never say only "Students had fun." Always explain WHY it matters.

------------------------------------------
5 PARENT EMOTION
Always include one sentence that makes parents imagine their own child.
(e.g., Today, our students weren't just selling products. They were discovering confidence one customer at a time.)
When parents finish reading, they should feel "My child would love to learn here."
Never make the school the hero. Students are always the hero.

------------------------------------------
6 BRAND PROMISE
Must include one phrase like: Learning Beyond the Classroom, Experience Creates Confidence, Growing Curious Minds, Preparing Future Leaders, Discover, Create, Grow, Every Child Matters.

------------------------------------------
7 INVITATION (CTA)
Invite parents. (e.g. Come experience learning beyond the classroom. / Admissions are now open.)

------------------------------------------
8 CONTACT
  -ภาษาอังกฤษ:
  Contact us
  Call (+66) 38 611 251
  Email: mail@somkidvittaya.ac.th
  Website: somkidvittaya.ac.th
  School visit: https://calendar.app.google/HhhN11dAj8r3HehM7

  -ภาษาจีน:
  联系我们
  电话: (+66) 38 611 251
  电子邮箱: mail@somkidvittaya.ac.th
  官方网站: somkidvittaya.ac.th
  预约参观学校: https://calendar.app.google/HhhN11dAj8r3HehM7

  -ภาษาไทย:
  ติดต่อเรา
  โทรศัพท์: (+66) 38 611 251
  อีเมล: mail@somkidvittaya.ac.th
  เว็บไซต์: somkidvittaya.ac.th
  นัดหมายเยี่ยมชมโรงเรียน: https://calendar.app.google/HhhN11dAj8r3HehM7

  • ชื่อโรงเรียน: "โรงเรียนสมคิดวิทยา" / "Somkidvittaya School" / "somkidvittaya学校" เท่านั้น
  • ระดับชั้นภาษาอังกฤษและจีนใช้ G.แทนประถม และ K.แทนอนุบาล ส่วนภาษาไทยใช้ป.แทนประถม และ อ.แทนอนุบาล

------------------------------------------
9 HASHTAGS
Include about 10 hashtags (Brand: 3, Learning: 2, Activity: 3, Future Skills: 2). DO NOT REPEAT. Do not write hashtags directly in the text body. Put them at the very end.

------------------------------------------
10 SIGNATURE ENDING
Every post should end with a thematic signature (e.g. "Every journey begins with curiosity.", "The future starts here.")

==================================================
WRITING STYLE
==================================================
Tone: Premium, Warm, Confident, International, Professional, Optimistic
Avoid: จัดกิจกรรม, เพื่อส่งเสริม, บรรยากาศเต็มไปด้วย, นักเรียนได้มีโอกาส, นักเรียนได้ร่วม
Instead: Describe moments. Describe transformation.
Preferred Vocabulary: Experience, Discover, Explore, Create, Grow, Future Ready, Hands-on Learning, Meaningful Learning, Confidence, Leadership, Innovation, Creativity, Collaboration, Curiosity, Character
Avoid repeating phrases used in previous sections. Every paragraph should introduce new information.

==================================================
HERO QUOTE
==================================================
Generate one inspirational quote. Maximum 12 words. No punctuation at end. Can be used on cover artwork. Return hero_quote inside JSON.

==================================================
SELF QUALITY CHECK
==================================================
Before returning JSON, Check:
✓ Does this sound like an international school?
✓ Would parents feel proud?
✓ Is student growth obvious?
✓ Is there emotional storytelling?
✓ Does it strengthen the Somkidvittaya brand?
If not, rewrite before returning JSON.

Evaluate and return caption_score (0-100). If score < 90, rewrite once internally.
`;

    var rawResponse = callGeminiAPI(base64ImagesArray, mimeType, prompt);'''

content = prompt_regex.sub(new_prompt, content)


# 2. Replace JSON Schema
schema_regex = re.compile(r'var jsonSchema = \{.*?\}\s*;\s*var partsArray', re.DOTALL)
new_schema = r'''var jsonSchema = {
    "type": "OBJECT",
    "properties": {
      "kept_image_indices": {
        "type": "ARRAY",
        "description": "Index ของรูปที่เก็บไว้ รูปแรกต้องเป็นภาพปกที่คมชัดที่สุด",
        "items": { "type": "INTEGER" }
      },
      "cover_headline": {
        "type": "OBJECT",
        "properties": {
          "headline": { "type": "STRING" },
          "subhead":  { "type": "STRING" },
          "detail":   { "type": "STRING" }
        },
        "required": ["headline", "subhead", "detail"]
      },
      "cover_design": {
        "type": "OBJECT",
        "properties": {
          "mood": { "type": "STRING" },
          "headline": { "type": "STRING" },
          "subhead": { "type": "STRING" },
          "detail": { "type": "STRING" },
          "primary_color": { "type": "STRING" },
          "secondary_color": { "type": "STRING" },
          "focus": { "type": "STRING" }
        }
      },
      "hero_quote": { "type": "STRING" },
      "caption_score": { "type": "INTEGER" },
      "post_caption": {
        "type": "OBJECT",
        "properties": {
          "facebook": {
            "type": "OBJECT",
            "properties": {
              "english": { "type": "STRING" },
              "chinese": { "type": "STRING" },
              "thai": { "type": "STRING" }
            },
            "required": ["english", "chinese", "thai"]
          },
          "instagram": {
            "type": "OBJECT",
            "properties": {
              "english": { "type": "STRING" },
              "chinese": { "type": "STRING" },
              "thai": { "type": "STRING" }
            },
            "required": ["english", "chinese", "thai"]
          }
        },
        "required": ["facebook", "instagram"]
      }
    },
    "required": ["kept_image_indices", "cover_headline", "post_caption"]
  };

  var partsArray'''

content = schema_regex.sub(new_schema, content)


# 3. Replace generationConfig & add systemInstruction
payload_regex = re.compile(r'var payload = \{\s*"contents": \[\{ "parts": partsArray \}\],\s*"generationConfig": \{\s*"responseMimeType": "application/json",\s*"responseSchema": jsonSchema\s*\}\s*\};', re.DOTALL)
new_payload = r'''var payload = {
    "contents": [{ "parts": partsArray }],
    "systemInstruction": {
      "parts": [{
        "text": "Always think before writing.\nFirst, analyze the images.\nThen, identify the educational value.\nThen, connect to Somkidvittaya's philosophy.\nFinally, write naturally.\nNever expose your reasoning."
      }]
    },
    "generationConfig": {
      "responseMimeType": "application/json",
      "responseSchema": jsonSchema,
      "temperature": 0.85,
      "topP": 0.95,
      "topK": 40,
      "maxOutputTokens": 4096
    }
  };'''

content = payload_regex.sub(new_payload, content)

with open('Code.gs', 'w') as f:
    f.write(content)

File: patch.js
const fs = require('fs');
let codeGs = fs.readFileSync('Code.gs', 'utf8');

// Replace doPost top part to support JSON payload
codeGs = codeGs.replace(/function doPost\(e\) \{\n  try \{\n    var action = e\.parameter\.action;/, 
`function doPost(e) {
  try {
    var params = e.parameter;
    if (e.postData && e.postData.contents) {
      try {
        var parsed = JSON.parse(e.postData.contents);
        params = Object.assign({}, params, parsed);
      } catch(err) {}
    }
    var action = params.action;`);

// Replace e.parameter with params in doPost
codeGs = codeGs.replace(/e\.parameter/g, 'params');

// But handlePublishToSocial is a separate function, let's fix it too
codeGs = codeGs.replace(/function handlePublishToSocial\(e\)/, 'function handlePublishToSocial(params)');

// In handlePublishToSocial, replace e.parameter with params
codeGs = codeGs.replace(/e\.parameter/g, 'params');

// Fix handleTranslateCaption
codeGs = codeGs.replace(/function handleTranslateCaption\(e\)/, 'function handleTranslateCaption(params)');

// Fix Video Handlers
codeGs = codeGs.replace(/function handleVideoStepFB\(e\)/, 'function handleVideoStepFB(params)');
codeGs = codeGs.replace(/function handleVideoStepIG\(e\)/, 'function handleVideoStepIG(params)');
codeGs = codeGs.replace(/function handleVideoCheckIG\(e\)/, 'function handleVideoCheckIG(params)');
codeGs = codeGs.replace(/function handleVideoPublishIG\(e\)/, 'function handleVideoPublishIG(params)');

fs.writeFileSync('Code.gs', codeGs);

let scriptJs = fs.readFileSync('public/script.js', 'utf8');

// Replace processPost fetch
scriptJs = scriptJs.replace(/body: new URLSearchParams\(\{\n                'images': JSON\.stringify\(imagesDataForAI\),\n                'mimeType': 'image\/jpeg',\n                'activityInfo': activityContext,\n                'mediaMode': mediaMode\n            \}\)/, 
`body: JSON.stringify({
                'images': JSON.stringify(imagesDataForAI),
                'mimeType': 'image/jpeg',
                'activityInfo': activityContext,
                'mediaMode': mediaMode
            })`);
scriptJs = scriptJs.replace(/headers: \{ 'Content-Type': 'application\/x-www-form-urlencoded' \}/, 
`headers: { 'Content-Type': 'text/plain;charset=utf-8' }`);

// Replace confirmPublish fetch
scriptJs = scriptJs.replace(/requestBody = new URLSearchParams\(\{\n                'action': 'publishToSocial',\n                'mediaMode': 'photo',\n                'fbImages': JSON\.stringify\(processedImages\.fbImages\),\n                'igImages': JSON\.stringify\(processedImages\.igImages\),\n                'fbCaption': fbCaption,\n                'igCaption': igCaption\n            \}\);/g, 
`requestBody = JSON.stringify({
                'action': 'publishToSocial',
                'mediaMode': 'photo',
                'fbImages': JSON.stringify(processedImages.fbImages),
                'igImages': JSON.stringify(processedImages.igImages),
                'fbCaption': fbCaption,
                'igCaption': igCaption
            });`);
// There are multiple headers: { 'Content-Type': 'application/x-www-form-urlencoded' } in script.js, let's just replace all
scriptJs = scriptJs.replace(/headers: \{ 'Content-Type': 'application\/x-www-form-urlencoded' \}/g, 
`headers: { 'Content-Type': 'text/plain;charset=utf-8' }`);

// Also fix callStep inside video mode
scriptJs = scriptJs.replace(/body: new URLSearchParams\(params\)/g, `body: JSON.stringify(params)`);

fs.writeFileSync('public/script.js', scriptJs);

console.log("Patched!");
